backend/cellguide/pipeline/source_collections/source_collections_generator.py

```diff
@@ -0,0 +1,77 @@
+    """
+    all_datasets = get_datasets_from_curation_api()
+    all_collections = get_collections_from_curation_api()
+
+    collections_dict = {collection["collection_id"]: collection for collection in all_collections}
+    datasets_dict = {dataset["dataset_id"]: dataset for dataset in all_datasets}
+
+    source_collections_data: dict[str, list[SourceCollectionsData]] = {}
+
+    for cell_id in all_cell_type_ids_in_corpus:
+        lineage = descendants(cell_id)
+        assert cell_id in lineage, f"{cell_id} not found in lineage"
+
+        # Generate a list of unique dataset ids that contain cell type ids in the lineage
+        dataset_ids: list[str] = []
+        for dataset_id in datasets_dict:
+            for cell_type in datasets_dict[dataset_id]["cell_type"]:
+                if cell_type["ontology_term_id"] in lineage:
+                    dataset_ids.append(dataset_id)
+                    break
+        assert len(set(dataset_ids)) == len(dataset_ids)
+
+        # Generate a mapping from collection id to SourceCollectionsData
+        collections_to_source_data = {}
+        for dataset_id in dataset_ids:
+            dataset = datasets_dict[dataset_id]
+            collection_id = dataset["collection_id"]
+
+            # If we don't have any source data on this collection yet, create a new SourceCollectionsData object
+            if collection_id not in collections_to_source_data:
+                collection = collections_dict.get(collection_id)
+                source_data = SourceCollectionsData(
+                    collection_name=collection["name"],
+                    collection_url=collection["collection_url"],
+                    publication_url=collection["doi"],
+                    publication_title=format_citation(collection["publisher_metadata"])
+                    if collection["publisher_metadata"]
+                    else "No publication",
+                    tissue=[],
+                    disease=[],
+                    organism=[],
+                )
+                collections_to_source_data[collection_id] = source_data
+
+            # Add the tissue, disease, and organism metadata from the dataset to the SourceCollectionsData object. If we
+            # previously found a SourceCollectionsData object, we'll want to add the tissue/disease/organism from this
+            # dataset to the existing object. If not, we'll want to create a new list for each of these fields.
+            source_data = collections_to_source_data[collection_id]
+
+            for tissue in dataset["tissue"]:
+                if tissue["ontology_term_id"] not in [t["ontology_term_id"] for t in source_data.tissue]:
+                    source_data.tissue.append(tissue)
+
+            for disease in dataset["disease"]:
+                if disease["ontology_term_id"] not in [d["ontology_term_id"] for d in source_data.disease]:
+                    source_data.disease.append(disease)
```

Comment:
out of curiosity, have you observed duplicate diseases/tissues/organisms in the source data or is this just a guardrail?

Reply:
i have not - this is just a slightly rewritten guardrail from what was originally in the python notebook